include/unifex/detail/any_heap_allocated_storage.hpp

```diff
@@ -0,0 +1,164 @@
+                std::allocator_arg_t,
+                allocator_type allocator,
+                std::in_place_type_t<T>,
+                Args&&... args)
+          : object(static_cast<Args&&>(args)...)
+          , allocator(std::move(allocator)) {}
+
+        UNIFEX_NO_UNIQUE_ADDRESS T object;
+        UNIFEX_NO_UNIQUE_ADDRESS allocator_type allocator;
+      };
+
+      // This is the base-class object that holds the pointer to the
+      // heap-allocated state.
+      class base {
+      public:
+        template(typename... Args)                     //
+            (requires constructible_from<T, Args...>)  //
+            explicit base(
+                std::allocator_arg_t,
+                allocator_type allocator,
+                std::in_place_type_t<T>,
+                Args&&... args) {
+          state_ = allocator_traits::allocate(allocator, 1);
+
+          auto deallocateOnExit = scope_guard{[&]() noexcept {
+            allocator_traits::deallocate(allocator, state_, 1);
+          }};
+
+          allocator_traits::construct(
+              allocator,
+              state_,
+              std::allocator_arg,
+              allocator,
+              std::in_place_type<T>,
+              static_cast<Args&&>(args)...);
+
+          deallocateOnExit.release();
+        }
+
+        template(typename... Args)                   //
+            (requires                                //
+             constructible_from<T, Args...> AND      //
+             (!same_as<Allocator, allocator_type>))  //
+            explicit base(
+                std::allocator_arg_t,
+                Allocator allocator,
+                std::in_place_type_t<T>,
+                Args&&... args)
+          : base(
+                std::allocator_arg,
+                allocator_type(std::move(allocator)),
+                std::in_place_type<T>,
+                static_cast<Args&&>(args)...) {}
+
+        // Ideally this would only be instantiated if it's actually called.
+        //
+        // This should have a 'requires std::copy_constructible<T>' on the
+        // definition but we can't currently do that with the concepts emulation
+        // layer so we'll just enable the check when we actually have concepts.
+        base(const base& other)
+#if UNIFEX_CXX_CONCEPTS
+            requires copy_constructible<T>
+#endif
+          : base(
+                std::allocator_arg,
+                const_cast<const allocator_type&>(other.state_->allocator),
+                std::in_place_type<T>,
+                const_cast<const T&>(other.state_->object)) {
+        }
+
+        base(base&& other) noexcept
+          : state_(std::exchange(other.state_, nullptr)) {}
+
+        ~base() {
+          if (state_ != nullptr) {
+            allocator_type allocCopy = std::move(state_->allocator);
+            state_->~state();
+            std::allocator_traits<allocator_type>::deallocate(
+                allocCopy, state_, 1);
+          }
+        }
```

Comment:
This type doesn't need move assignment - if the type erasing wrapper wants to implement assignment then it will usually destroy the current value and construct a new one - it won't call through to the underlying move-assignment operator of the concrete type.

Reply:
That's fine, except that I wouldn't implement it the way you describe. That has the potential to leave the object in an invalid state, and violates invariants in the case of self-move. Instead, I would:

```c++
base& operator=(base b) {
  swap(b);
  return *this;
}
```

Obvs, you need to implement `swap` as well.

Reply:
Self-move-assignment should be handled at the `any_object::operator=()` implementation.
It would need to handle both the inline/heap storage strategies anyway.

e.g.
```c++
any_object& operator=(any_object&& other) noexcept(RequireNothrowMove) {
  if (this != &other) {
    vtable_->template get<destroy_t>()(&storage_);
    vtable_ = empty_vtable();
    other.vtable_->template get<move_construct_t>()(&storage_, &other.storage_);
    vtable_ = other.vtable_;
  }
  return *this;
}
```